### ISSUE TITLE
ValueError when re-compiling function with a multi-dimensional array as a static field

### ISSUE THREAD

**Issue by bheijden** [jax-ml/jax#24204]
### Description

A `ValueError` occurs when re-compiling a function with a re-instantiated  `flax.struct.dataclass` with a **static** and **multi-dimensional** field . The issue seems related to how static equality is checked for cached compiled functions, as it happens only when setting the `pytree_node` to `False` in the dataclass field. The error is triggered on the second iteration of the loop when the `jax.jit` compiled function is called a **second time** with a non-scalar `jax.Array`.

The error message:
```
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

This error does not occur when:
- The field is set as a scalar (`jax.numpy.array(1)`).
- The `pytree_node` is set to `True`.
- The parameters are **not** re-instantiated within the loop.

**Minimal Working Example:**

```python
import jax
from flax import struct

@struct.dataclass
class Foo:
    # Error has to do with this field when set as static.
    bar: jax.Array = struct.field(pytree_node=False)  # If True, the error will not occur

def step(_params):
    return _params

params = Foo(bar=jax.numpy.arange(4))
for i in range(2):
    print(f"i: {i}")  # Fails on second iteration

    # When we re-instantiate, but with a scalar, the error will not occur
    # params = Foo(bar=jax.numpy.array(1))  # Works, because foo.bar.all() is valid I guess

    # Jit the step function
    jit_step = jax.jit(step).lower(params).compile()
    # ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

    # Only when re-instantiating the dataclass, the error will occur
    params = Foo(bar=jax.numpy.arange(4))
```

**Expected Behavior:**
I expect `jax` to either throw an error on the first iteration, similar to when non-hashable arrays are provided as static args, or not throw an error at all and function correctly.

### System info (python version, jaxlib version, accelerator, etc.)

```
flax:   0.8.5
jax:    0.4.30
jaxlib: 0.4.30
numpy:  1.26.4
python: 3.9.19 (main, Apr  6 2024, 17:57:55)  [GCC 9.4.0]
jax.devices (1 total, 1 local): [cuda(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='r2ci-Alienware-m15-R4', release='5.15.0-122-generic', version='#132~20.04.1-Ubuntu SMP Fri Aug 30 15:50:07 UTC 2024', machine='x86_64')
$ nvidia-smi
Wed Oct  9 12:10:13 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 3070 ...    Off | 00000000:01:00.0  On |                  N/A |
| N/A   57C    P0              45W / 125W |    816MiB /  8192MiB |      1%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A      1600      G   /usr/lib/xorg/Xorg                          139MiB |
|    0   N/A  N/A      1939      G   /usr/lib/xorg/Xorg                          199MiB |
|    0   N/A  N/A      2070      G   /usr/bin/gnome-shell                         99MiB |
|    0   N/A  N/A      6512      G   /usr/lib/firefox/firefox                    140MiB |
|    0   N/A  N/A      6763      G   ...seed-version=20241008-180117.502000       35MiB |
|    0   N/A  N/A      7480      G   ...ures=SpareRendererForSitePerProcess       25MiB |
|    0   N/A  N/A     11398      C   /home/r2ci/rex/.venv/bin/python             144MiB |
+---------------------------------------------------------------------------------------+
```

**Comment by bheijden** [comment]
Yes, I understand your explanation. My point is more about the user-facing consistency of the behavior. For users who are not familiar with the internal details, it feels inconsistent that the error doesn't occur on the first iteration if static arrays are not allowed. Furthermore, while arrays are apparently not permitted as static fields, scalar arrays (e.g., jax.numpy.array(1)) don't raise any errors, which adds to the confusion.

Finally, the error message itself suggests that a check for .all() in the case of multidimensional arrays might resolve the issue, potentially enabling static array definitions. It seems like there could be room for improvement here to make this behavior more intuitive or consistent for users who might not be aware of these internal mechanisms.

**Comment by jakevdp** [comment]
Thanks, that makes sense.

I'm not sure how flax or JAX could do better on this to be honest: attributes marked static are expected to have certain properties, but pre-emptively checking those properties is somewhat costly. For example, one way to do it would be for every time a pytree object is constructed, we go through the elements and check that `__eq__` and `__hash__` behave as expected. That would lead to the consistent errors you're asking for.

Unfortunately, pytrees are constructed *very* frequently, so adding such checks would add a lot of overhead to every JAX program. The alternative is to only check the required properties when they're used: this leads to better performance, but as you note is less consistent from the user perspective.

I think the performance piece is critical enough that the current behavior is a better choice – what do you think?

**Comment by bheijden** [comment]
I agree that adding checks to every pytree creation would introduce a significant performance overhead that likely wouldn't justify the benefit.

However, would it be possible to validate static arguments during the cache-hit check instead? Since the traversal is already happening at that point, adding a check there would only impact cases where a previously jitted function is called, which should be less frequent compared to pytree creation. This way, a more descriptive error could be raised—rather than failing with a .all() check—while still allowing scalar arrays to pass through without issue.

Lastly, I was also surprised that fields marked as "static" are still traversed like pytrees. I had expected them to be treated as static leaves and not further traversed. In such cases, could we allow users to override __hash__ and __eq__ for those fields, making them valid for caching?

**Comment by jakevdp** [comment]
I don't think the static leaves are being further traversed – what's your evidence of that?

**Comment by bheijden** [comment]
Well, it seemed that way when I placed a breakpoint [here](https://github.com/jax-ml/jax/blob/9cf952a535518da59cdcecc9145dba287beddca2/jax/_src/array.py#L289) and modified the `Foo` dataclass to contain a nested `Bar` dataclass. I wanted to check how often the code would hit the breakpoint.

In this updated example, I encountered the breakpoint twice—once for each field (`b` and `c`) in `Bar`—even though only the multi-dimensional array (`c`) caused the error.

```python
import jax
from flax import struct

@struct.dataclass
class Bar:
    b: jax.Array  # Scalar array
    c: jax.Array  # Multi-dimensional array

@struct.dataclass
class Foo:
    a: Bar = struct.field(pytree_node=False)  # Static field

def step(_params):
    return _params

params = Foo(a=Bar(b=jax.numpy.array(1), c=jax.numpy.arange(4)))
for i in range(2):
    print(f"i: {i}")  # Fails on the second iteration

    # Jit the step function
    jit_step = jax.jit(step).lower(params).compile()
    # ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

    # Re-instantiating the dataclass triggers the error on the second iteration
    params = Foo(a=Bar(b=jax.numpy.array(1), c=jax.numpy.arange(4)))
```

This suggests that even though `pytree_node=False` is set for `a`, the fields within `Bar` (`b` and `c`) are still being traversed, at least during the cache-hit process. This traversal only causes an issue when encountering the multi-dimensional array (`c`), but both fields are processed.

**Comment by jakevdp** [comment]
Thanks for the example – agreed that is strange: I would also expect `Bar` to be treated as a static variable, and not traversed by the pytree machinery.

**Comment by bheijden** [comment]
I believe the traversing happens because of how dataclasses implement the `__eq__` function, where an element-wise comparison occurs, similar to how tuples are checked. This likely gives the impression that the `pytree` is being traversed, even though it's a result of the equality check.

Regarding the immutability and hashability concerns, I was under the impression that Flax dataclasses are immutable by default and generate a hash function, since `frozen=True` is set [here](https://github.com/google/flax/blob/53ba626c9396aeed31d9c038616214f5e5e764d4/flax/struct.py#L110).

According to the [Python documentation](https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass) on dataclasses, when both `eq=True` and `frozen=True` are set (which is the default behavior in Flax), a hash function is automatically generated, even if `unsafe_hash=False`.

To summarize from the docs:

- When `eq=True` and `frozen=True` (as in the default Flax dataclass setup), a `__hash__()` method is generated.
- This makes the class instances hashable, ensuring immutability.
- If `frozen=False`, no `__hash__()` method is generated, making the class unhashable.

However, I'm not entirely sure what is meant by a "meaningful" hash in this context. From what I understand, the hash is based on the `__eq__` function, which compares all elements of the dataclass fields. To me, this seems like a reasonable and consistent approach for determining equality and hashability.

**Comment by jakevdp** [comment]
> Regarding the immutability and hashability concerns, I was under the impression that Flax dataclasses are immutable by default and generate a hash function, since `frozen=True` is set [here](https://github.com/google/flax/blob/53ba626c9396aeed31d9c038616214f5e5e764d4/flax/struct.py#L110).

Ah, I see, thanks. That said, this does not mean it's OK to mark a non-hashable attribute as static. That's the source of your problem: we should probably improve the error here to make that more clear, but the only fix here is for you to not mark arrays as static.

**Comment by jakevdp** [comment]
Here's a more stripped-down repro:
```python
import jax
from dataclasses import dataclass

@dataclass
class Foo:
  a: jax.Array
jax.tree_util.register_dataclass(Foo, data_fields=[], meta_fields=['a'])

func = jax.jit(lambda x: x)
foo1 = Foo(jax.numpy.arange(4))
func(foo1)  # no error

foo2 = Foo(jax.numpy.arange(4))
func(foo2)  # error during check of foo1 == foo2
```
In this case, `func(foo1)` should error because of non-hashable static attributes, and I'm not sure why that is not happening.

**Comment by jakevdp** [comment]
Maybe even more to the point, the issue is that this hash computation should fail, because the tree structure contains an unhashable array.
```python
import jax
from dataclasses import dataclass

@dataclass
class Foo:
  a: jax.Array
jax.tree_util.register_dataclass(Foo, data_fields=[], meta_fields=['a'])

foo = Foo(jax.numpy.arange(4))
foo_struct = jax.tree.structure(foo)
hash(foo_struct)  # should error but doesn't
```

**Comment by jakevdp** [comment]
> If it's an expected error, I would have expected it to happen on the first iteration. However, it only happend in the second iteration, which is unexpected.

Chatting with @hawkinsp, it sounds like this is the expected behavior. Hash functions do not necessarily reference all objects in a structure, and so it's not guaranteed that hashing a pytree structure will always identify a problematic value in that structure. But on the second call, `__eq__` is called, and this will *always* fail for arrays that are marked as static.

I'm wondering how to best document this – it's not clear to me. What do you think?

**Comment by bheijden** [comment]
#### Problem Description

To clarify, it appears that the unhashability of `jax` arrays is a deliberate design choice. Despite the immutability of `jax` arrays, which theoretically could allow hashing, hashing large arrays could be computationally expensive and undesirable, especially in scenarios like repeated `jit` calls. Is this reasoning correct?

The primary issue seems to stem from how the `__eq__` method operates for both `numpy` and `jax` arrays. When comparing these arrays, `__eq__` does not return a single boolean value, but rather an array of booleans. This becomes problematic in the context of `flax.struct.dataclass`, which evaluates equality by comparing the fields of the class, effectively treating it as a tuple of its fields. Since Python cannot interpret an array of booleans as a single truth value, the equality check raises an error when it encounters an array. 

#### Potential Solution

Could an initial call to the `__eq__` method address this issue? For example, calling `__eq__` with itself on the first iteration could trigger a consistent failure if any field returns a non-boolean result. This might help surface the error more predictably and ensure it fails early, filtering out unsupported types (like arrays) from being used as fields of a statically marked argument.

#### Documentation Feedback

In terms of documentation, this behavior is not immediately clear. When using a `flax.struct.dataclass`, I expected that the dataclass would handle hashing and equality comparisons out of the box. However, it's not obvious that each field within the dataclass must also have a valid equality operator that returns a single boolean value. Not sure if this is something that must be documented in `flax`, or in `jax`...

**Comment by jakevdp** [comment]
> To clarify, it appears that the unhashability of `jax` arrays is a deliberate design choice. Despite the immutability of `jax` arrays, which theoretically could allow hashing, hashing large arrays could be computationally expensive and undesirable, especially in scenarios like repeated `jit` calls. Is this reasoning correct?

Yes – but also hashability in Python requires particular `__eq__` semantics that JAX arrays do not follow: rather than `arr1 == arr2` returning a boolean, JAX follows NumPy in returning an array of element-wise comparisons. So we must set `__hash__ = None` because of how we define `__eq__` (see the [Python `__hash__` docs](https://docs.python.org/3.5/reference/datamodel.html#object.__hash__) for more).

> Could an initial call to the __eq__ method address this issue? For example, calling __eq__ with itself on the first iteration could trigger a consistent failure if any field returns a non-boolean result. 

Maybe, but note that just calling `__eq__` between a pytree struct and itself would not be enough. For example, this does not error:
```python
import jax
from dataclasses import dataclass

@dataclass
class Foo:
  a: jax.Array
jax.tree_util.register_dataclass(Foo, data_fields=[], meta_fields=['a'])

foo = Foo(jax.numpy.arange(4))
foo_struct = jax.tree.structure(foo)
foo_struct == foo_struct  # True
```
This is due to the implementation of `__eq__` for Python tuples, which won't call `__eq__` on their contents if the object IDs match. So it's not a solution here, and even if it were, it's not clear that adding that overhead to every JIT dispatch would be worth the cost.

> In terms of documentation, this behavior is not immediately clear. When using a flax.struct.dataclass, I expected that the dataclass would handle hashing and equality comparisons out of the box. However, it's not obvious that each field within the dataclass must also have a valid equality operator that returns a single boolean value. Not sure if this is something that must be documented in flax, or in jax...

The JAX docs are pretty clear that arrays cannot be marked as static. For example, from the docs for [`jax.tree_util.register_dataclass`](https://jax.readthedocs.io/en/latest/_autosummary/jax.tree_util.register_dataclass.html), which `flax.struct.dataclass` uses internally:

> meta_fields *(Sequence[[str](https://docs.python.org/3/library/stdtypes.html#str)])* – auxiliary data field names. These fields *must* contain static, hashable, immutable objects, as these objects are used to generate JIT cache keys. In particular, `meta_fields` cannot contain [`jax.Array`](https://jax.readthedocs.io/en/latest/_autosummary/jax.Array.html#jax.Array) or [`numpy.ndarray`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray) objects.

Maybe the flax docs should do more to warn users about this?

**Comment by jakevdp** [comment]
Revisiting this – I can't think of any way to surface this error earlier that does not lead to dispatch overhead that we want to avoid. Do you have any thoughts?

**Comment by bheijden** [comment]
Let me break down my thoughts based on your points, focusing on the steps where the error arises and a potential (non ideal) solution.

### Steps to Reproduce the Issue

In my original minimal working example (MWE), I distinguish three distinct steps in the process:

1. **Lowering and Compilation**: The function is lowered and compiled using a given pytree input.
2. **First Execution (Dispatch)**: The JIT-compiled function is executed for the first time using the *same* pytree input as in step 1.
3. **Subsequent Execution**: The JIT-compiled function is called again, but this time with a pytree that has a similar structure but different object IDs.

Currently, the error is only raised in step 3 when a static argument (e.g., a `jax.Array`) with a non-boolean-returning `__eq__` is encountered. 

### Problem

Ideally, the error should be raised earlier—during steps 1 or 2. However:

- **Step 2 (Execution)**: Adding a check at this step could significantly increase overhead because it would need to run every time the compiled function is called, which seems undesirable.
- **Step 1 (Compilation)**: It seems more appropriate for the check to occur here, as this step only happens once during the lifetime of the compiled function.

From what I understand about JAX's caching behavior, each JIT-compiled function maintains an LRU cache of compiled functions, where the cache key is determined by:
- The shapes and dtypes of dynamic arguments, and 
- The hash of static arguments.

### Proposed (non-ideal) Solution

Would it make sense to perform an explicit `__eq__` check on all static arguments during the compilation (step 1)? Specifically:
- Call the `__eq__` method on each static argument with a (deep?) copy of itself (to ensure a different object ID).
- Verify that the `__eq__` method returns both:
  - A single boolean value, and
  - `True` when compared to itself.

This would ensure that static arguments have the correct equality semantics and avoid unexpected failures during later steps. Additionally, because this check is only performed during compilation, it wouldn't introduce runtime overhead during subsequent calls.

### Considerations

1. **Deep Copy**: Creating a deep copy of static arguments could potentially be costly, especially for large data structures, and in some cases may raise errors... Would a shallow copy suffice in such cases? Probably not, because the problem may then simply pop-up a few levels deeper into the nested static object...
2. **Performance Trade-offs**: While this solution avoids runtime overhead, it does introduce additional checks during compilation. For workflows that frequently recompile functions, this could impact performance.
3. From the current implementation, is there any scenario where a static argument's `__eq__` method returning `False` when compared to itself is desirable or intended behavior? We could also only check whether a boolean is returned instead of the actual truth value.

**Comment by jakevdp** [comment]
> Would it make sense to perform an explicit `__eq__` check on all static arguments during the compilation (step 1)? Specifically:

So I responded to this in https://github.com/jax-ml/jax/issues/24204#issuecomment-2460404257; unfortunately due to the implementation of `__eq__`, it's not trivial to construct an `__eq__` check that will that will catch this error early.

Also, pre-emptively doing this kind of `__eq__` check would be a non-negligible cost in many situations, and we can't pay that cost on every JIT dispatch that includes a pytree argument.

I appreciate your re-outlining the issue each time for clarity, but I think the answer is still "there's no way to do this without sacrificing performance that we're not willing to sacrifice."

**Comment by jongyaoY** [comment]
In case you are still looking for solutions for setting ndarray as static field in a dataclass, here's how they do it in [Mujoco XLA](https://github.com/google-deepmind/mujoco/blob/main/mjx/mujoco/mjx/_src/dataclasses.py#L37) that will solve ndarray being non-hashable issue

```python
def dataclass(clz: _T, register_as_pytree: bool) -> _T:
  """Wraps a dataclass with metadata for which fields are pytrees.

  This is based off flax.struct.dataclass, but instead of using field
  descriptors to specify which fields are pytrees, we follow a simple rule:
  a leaf field is a pytree node if and only if it's a jax.Array

  Args:
    clz: the class to register as a dataclass

  Returns:
    the resulting dataclass, registered with Jax
  """
  data_clz = dataclasses.dataclass(frozen=True)(clz)
  data_clz.replace = dataclasses.replace

  if register_as_pytree:
    meta_fields, data_fields = [], []
    for field in dataclasses.fields(data_clz):
      if _jax_in_args(field.type):
        data_fields.append(field)
      else:
        meta_fields.append(field)

    def iterate_clz_with_keys(x):
      def to_meta(field, obj):
        val = getattr(obj, field.name)
        # numpy arrays are not hashable so return raw bytes instead
        if isinstance(val, np.ndarray):
          return (val.tobytes(), val.dtype, val.shape)
        else:
          return val

      def to_data(field, obj):
        return (jax.tree_util.GetAttrKey(field.name), getattr(obj, field.name))

      data = tuple(to_data(f, x) for f in data_fields)
      meta = tuple(to_meta(f, x) for f in meta_fields)
      return data, meta

    def clz_from_iterable(meta, data):

      def from_meta(field, meta):
        if field.type is np.ndarray:
          arr = np.frombuffer(meta[0], dtype=meta[1]).reshape(meta[2])
          return (field.name, arr)
        else:
          return (field.name, meta)

      from_data = lambda field, meta: (field.name, meta)

      meta_args = tuple(from_meta(f, m) for f, m in zip(meta_fields, meta))
      data_args = tuple(from_data(f, m) for f, m in zip(data_fields, data))

      return data_clz(**dict(meta_args + data_args))

    jax.tree_util.register_pytree_with_keys(
        data_clz, iterate_clz_with_keys, clz_from_iterable
    )

  return data_clz
```

with this dataclass you can just specify the static fields as np.ndarray

**Comment by jakevdp** [comment]
Thanks for the suggestion @jongyaoY – you didn't define `_jax_in_args`, so it's hard to tell exactly what this snippet is doing. But it looks like it may be dynamically detecting whether an attribute should be a data field or meta field: please note that this kind of dynamic lowering approach will fail in many cases, e.g. in `vmap` which uses non-array sentinel objects to detect the pytree structure. To address this, you should switch to a solution where the data and meta arguments are declared statically, and not determined dynamically based on the contents of the particular object being flattened.

**Comment by jongyaoY** [comment]
> Thanks for the suggestion [@jongyaoY](https://github.com/jongyaoY) – you didn't define `_jax_in_args`, so it's hard to tell exactly what this snippet is doing. But it looks like it may be dynamically detecting whether an attribute should be a data field or meta field: please note that this kind of dynamic flattening approach will fail in many cases. For example, `vmap` uses non-array sentinel objects to detect the pytree structure. To address this, you should switch to a solution where the data and meta arguments are declared statically, and not determined dynamically based on the contents of the particular object being flattened.

Thanks for pointing that out; I've copied pasted the missing code block too. Again, the code is from [Mujoco XLA](https://github.com/google-deepmind/mujoco/blob/main/mjx/mujoco/mjx/_src/dataclasses.py#L37). As for the potential problems you mentioned, for example with `vmap`, so far I haven't had issues with passing the resulting dataclass to functions wrapped with `vmap` at least

**Comment by jakevdp** [comment]
Oh I see - you're marking things as data or meta fields based on the type annotation, not based on the runtime value. That should be fine then.

**Comment by jakevdp** [comment]
Closing because this is not actionable.
